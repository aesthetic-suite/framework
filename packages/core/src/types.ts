import {
  CustomProperties,
  GlobalStyleSheet,
  GlobalStyleSheetNeverize,
  LocalBlock,
  LocalStyleSheet,
  LocalStyleSheetNeverize,
  PropertyHandlerMap,
} from '@aesthetic/sss';
import { Utilities } from '@aesthetic/system';
import {
  ClassName,
  ColorScheme,
  ContrastLevel,
  Direction,
  DirectionConverter,
  ThemeName,
  Unit,
  UnitFactory,
  VendorPrefixer,
} from '@aesthetic/types';
import StyleSheet from './StyleSheet';

// Re-export for convenience
export { CustomProperties, GlobalStyleSheet, LocalBlock, LocalStyleSheet };

// And add aliases too
export type ElementStyles = LocalBlock;
export type ComponentStyles = LocalStyleSheet<ElementStyles>;
export type ThemeStyles = GlobalStyleSheet<ElementStyles>;

export type CompiledClassMap = Record<string, ClassName | ClassName[]>;

export interface CompiledRenderResult {
  result?: ClassName | CompiledClassMap;
  variants?: Record<string, ClassName | CompiledClassMap>;
  variantTypes?: string[];
}

export type CompiledRenderResultSheet = Record<
  string,
  ClassName | ClassName[] | CompiledRenderResult
>;

export interface RenderResult<T> {
  result?: T;
  variants?: Record<string, T>;
  variantTypes?: Set<string>;
}

export type RenderResultSheet<Result, Keys extends string = string> = {
  [K in Keys]?: RenderResult<Result>;
};

export interface SheetParams {
  contrast?: ContrastLevel;
  deterministic?: boolean;
  direction?: Direction;
  scheme?: ColorScheme;
  theme?: ThemeName;
  unit?: Unit;
  vendor?: boolean;
}

export interface SheetParamsExtended extends SheetParams {
  customProperties?: PropertyHandlerMap;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type BaseSheetFactory = (utils: Utilities<any>) => object;

export type GlobalSheetFactory<Shape = unknown, Block extends object = LocalBlock> = (
  utils: Utilities<Block>,
) => Shape extends unknown
  ? GlobalStyleSheet<Block>
  : GlobalStyleSheet<Block> & GlobalStyleSheetNeverize<Shape, Block>;

export type GlobalSheet<Shape, Block extends object, Result> = Omit<
  StyleSheet<Result, GlobalSheetFactory<Shape, Block>>,
  'addColorSchemeVariant' | 'addContrastVariant' | 'addThemeVariant'
>;

export type LocalSheetFactory<Shape = unknown, Block extends object = LocalBlock> = (
  utils: Utilities<Block>,
) => Shape extends unknown
  ? LocalStyleSheet<Block>
  : LocalStyleSheet<Block> & LocalStyleSheetNeverize<Shape, Block>;

export type LocalSheet<Shape, Block extends object, Result> = StyleSheet<
  Result,
  LocalSheetFactory<Shape, Block>
>;

export interface AestheticOptions {
  customProperties?: PropertyHandlerMap;
  defaultUnit?: Unit | UnitFactory;
  deterministicClasses?: boolean;
  directionConverter?: DirectionConverter | null;
  rootVariables?: boolean;
  vendorPrefixer?: VendorPrefixer | null;
}

export type EventType = 'change:direction' | 'change:theme';

export type EventListener = (...args: unknown[]) => void;

export type OnChangeDirection = (newDir: Direction) => void;

export type OnChangeTheme = (newTheme: ThemeName, results: unknown[]) => void;
