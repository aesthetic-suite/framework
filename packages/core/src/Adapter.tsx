import { isObject, stripClassPrefix } from 'aesthetic-utils';
import aesthetic from './instance';
import CacheManager from './CacheManager';
import Sheet from './Sheet';
import UnifiedSyntax from './UnifiedSyntax';
import { ClassName, CompiledStyleSheet, StyleName, TransformOptions } from './types';

export default abstract class Adapter<NativeBlock extends object> {
  protected cacheManager = new CacheManager<CompiledStyleSheet>();

  protected syntax = new UnifiedSyntax<NativeBlock>();

  adaptStyleSheet(nativeSheet: Sheet<NativeBlock>, styleName: StyleName): CompiledStyleSheet {
    return nativeSheet.toObject();
  }

  compileStyleSheet(styleName: StyleName, baseOptions?: TransformOptions): CompiledStyleSheet {
    const options = this.prepareTransformOptions(baseOptions);
    const cache = this.cacheManager.get(styleName, options);

    if (cache) {
      return cache;
    }

    // Apply global styles on first render
    this.applyGlobalStyles(baseOptions);

    // Convert unified syntax to adapter native syntax
    const nativeSheet = this.syntax.convertStyleSheet(
      aesthetic.getStyleSheet(styleName, options.theme),
      {
        ...options,
        name: styleName,
      },
    );

    // Parse and adapt native syntax
    const adaptedSheet = this.adaptStyleSheet(nativeSheet, styleName);

    return this.cacheManager.set(
      styleName,
      {
        ...adaptedSheet,
        ...nativeSheet.classNames,
      },
      options,
    );
  }

  /**
   * Transform the list of class names or style objects to a single class name.
   */
  transformStyles(styles: (undefined | false | ClassName | object)[]): ClassName {
    const classNames: ClassName[] = [];
    const objects: object[] = [];

    styles.forEach(style => {
      if (!style) {
        return;
      }

      if (typeof style === 'string') {
        classNames.push(
          ...String(style)
            .split(' ')
            .map(s => stripClassPrefix(s).trim()),
        );
      } else if (isObject(style)) {
        objects.push(style);
      } else if (__DEV__) {
        throw new Error('Unsupported style type to transform.');
      }
    });

    // Transform parsed blocks to class names
    if (objects.length > 0) {
      classNames.push(this.transformToClassName(objects));
    }

    return classNames.join(' ');
  }

  /**
   * Transform the compiled style objects into a class name.
   */
  abstract transformToClassName(styles: object[]): ClassName;

  /**
   * Return transform options with defaults applied.
   */
  protected prepareTransformOptions(options: TransformOptions = {}): Required<TransformOptions> {
    const dir = aesthetic.options.rtl ? 'rtl' : 'ltr';

    return {
      dir: options.dir || dir,
      global: options.global || false,
      name: options.name || '',
      theme: options.theme || aesthetic.options.theme,
    };
  }
}
