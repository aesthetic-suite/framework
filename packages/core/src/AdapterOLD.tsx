import deepMerge from 'extend';
import uuid from 'uuid/v4';
import { isObject, stripClassPrefix } from 'aesthetic-utils';
import CacheManager from './CacheManager';
import Sheet from './Sheet';
import StyleSheetManager from './StyleSheetManager';
import UnifiedSyntax from './UnifiedSyntax';
import { GLOBAL_STYLE_NAME } from './constants';
import {
  ClassName,
  TransformOptions,
  SheetMap,
  StyleName,
  StyleSheetFactory,
  ThemeName,
} from './types';

export default abstract class Adapter<
  Theme extends object,
  NativeBlock extends object,
  ParsedBlock extends object | string = NativeBlock
> {
  syntax = new UnifiedSyntax<NativeBlock>();

  protected sheetManager: StyleSheetManager | null = null;

  /**
   * Apply and inject global styles for the current theme.
   * This should only happen once!
   */
  applyGlobalStyles(baseOptions?: TransformOptions): this {
    // Templates may be generated by Webpack or something similar,
    // so set this programmatically based on the defined option.
    if (typeof document !== 'undefined') {
      document.documentElement.setAttribute('dir', this.options.rtl ? 'rtl' : 'ltr');
    }

    const options = this.getPreparedTransformOptions({
      ...baseOptions,
      global: true,
      name: GLOBAL_STYLE_NAME,
    });

    // Direction changes shouldn't regenerate global styles
    delete options.dir;

    const cache = this.cacheManager.get(GLOBAL_STYLE_NAME, options);
    const globalDef = this.globals[options.theme];

    if (cache || !globalDef) {
      return this;
    }

    const globalSheet = globalDef(this.getTheme(options.theme));
    const parsedSheet = this.cacheManager.set(
      GLOBAL_STYLE_NAME,
      this.parseStyleSheet(
        this.syntax.convertGlobalSheet(globalSheet, options).toObject(),
        GLOBAL_STYLE_NAME,
      ),
      options,
    );

    // Some adapters require the styles to be transformed to be flushed
    this.transformStyles(Object.values(parsedSheet), options);
    this.flushStyles(GLOBAL_STYLE_NAME);

    return this;
  }

  /**
   * Change the current theme to another registered theme.
   * This will purge all flushed global styles and regenerate new ones.
   */
  changeTheme(themeName: ThemeName): this {
    const oldTheme = this.options.theme;

    // Set theme as new option
    this.getTheme(themeName);
    this.options.theme = themeName;

    // Purge previous global styles
    this.purgeStyles(GLOBAL_STYLE_NAME);
    this.cacheManager.clear(unit => unit.global === true && unit.theme === oldTheme);

    // Generate new global styles
    this.applyGlobalStyles({ theme: themeName });

    return this;
  }

  /**
   * Create and return a style sheet unique to an adapter.
   */
  createStyleSheet(styleName: StyleName, baseOptions?: TransformOptions): SheetMap<ParsedBlock> {
    // Apply global styles on first render
    this.applyGlobalStyles(baseOptions);

    const nativeSheet = this.syntax.convertStyleSheet(
      this.getStyleSheet(styleName, options.theme),
      {
        ...options,
        name: styleName,
      },
    );

    const parsedSheet = this.parseStyleSheet(nativeSheet.toObject(), styleName);

    return this.cacheManager.set(
      styleName,
      {
        ...parsedSheet,
        ...nativeSheet.classNames,
      } as SheetMap<ParsedBlock>,
      options,
    );
  }

  /**
   * Compose and extend multiple style sheets to create 1 style sheet.
   */
  extendStyles(...styleSheets: StyleSheetFactory<Theme, any>[]): StyleSheetFactory<Theme, any> {
    return (theme: Theme) => {
      const sheets = styleSheets.map(sheet => sheet(theme));

      return deepMerge(true, {}, ...sheets);
    };
  }

  /**
   * Flush a target component's transformed styles and inject them into the DOM.
   * If no target defined, will flush all buffered styles.
   */
  flushStyles(styleName?: StyleName) {}

  /**
   * Return true if the style object is a parsed block and not a native block.
   */
  isParsedBlock(block: NativeBlock | ParsedBlock): block is ParsedBlock {
    return isObject(block);
  }

  /**
   * Parse an Aesthetic style sheet into an adapter native style sheet.
   */
  parseStyleSheet(styleSheet: SheetMap<NativeBlock>, styleName: StyleName): SheetMap<ParsedBlock> {
    // @ts-ignore Allow spread
    return { ...styleSheet };
  }

  /**
   * Purge and remove all styles from the DOM for the target component.
   * If no target defined, will purge all possible styles.
   */
  purgeStyles(styleName?: StyleName) {}

  /**
   * Return a native style sheet manager used for injecting CSS.
   */
  protected getStyleSheetManager(): StyleSheetManager {
    if (this.sheetManager) {
      return this.sheetManager;
    }

    this.sheetManager = new StyleSheetManager();

    return this.sheetManager;
  }
}
