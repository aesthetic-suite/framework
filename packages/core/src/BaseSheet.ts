export default abstract class BaseSheet {
  protected validateFactory<T>(factory: T): T {
    if (__DEV__) {
      const typeOf = typeof factory;

      if (typeOf !== 'function') {
        throw new TypeError(`A style sheet factory function is required, found "${typeOf}".`);
      }
    }

    return factory;
  }
}
