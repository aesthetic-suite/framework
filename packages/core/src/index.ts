/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Aesthetic } from './Aesthetic';
import { OverrideSheet } from './OverrideSheet';
import { Sheet } from './Sheet';

export * from './types';
export * from '@aesthetic/system';
export * from '@aesthetic/types';

export { Aesthetic, OverrideSheet, Sheet };
