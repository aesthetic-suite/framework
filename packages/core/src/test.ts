import { ThemeRegistry } from '@aesthetic/system';
import { darkTheme, design, lightTheme } from '@aesthetic/system/test';
import { ThemeName } from '@aesthetic/types';
import { Aesthetic, AestheticOptions, EventListener, EventType, GlobalSheet } from '.';

export { darkTheme, design, lightTheme };

export function resetAestheticState(aesthetic: Aesthetic<any, any>) {
	// @ts-expect-error Allow access
	aesthetic.activeDirection.reset();
	// @ts-expect-error Allow access
	aesthetic.activeTheme.reset();
	// @ts-expect-error Allow access
	aesthetic.listeners.clear();
	// @ts-expect-error Allow access
	aesthetic.globalSheetRegistry.clear();
	// @ts-expect-error Allow access
	aesthetic.themeRegistry.reset();

	aesthetic.configure({
		customProperties: {},
		defaultUnit: 'px',
		deterministicClasses: false,
		directionConverter: null,
		vendorPrefixer: null,
	});
}

export function getAestheticState(aesthetic: Aesthetic<any, any>): {
	activeDirection: string | undefined;
	activeTheme: string | undefined;
	globalSheetRegistry: Map<ThemeName, GlobalSheet<any, any, any>>;
	listeners: Map<EventType, Set<EventListener>>;
	options: AestheticOptions;
	themeRegistry: ThemeRegistry<any>;
} {
	return {
		// @ts-expect-error Allow access
		activeDirection: aesthetic.activeDirection.get(),
		// @ts-expect-error Allow access
		activeTheme: aesthetic.activeTheme.get(),
		// @ts-expect-error Allow access
		globalSheetRegistry: aesthetic.globalSheetRegistry,
		// @ts-expect-error Allow access
		listeners: aesthetic.listeners,
		// @ts-expect-error Allow access
		options: aesthetic.options,
		// @ts-expect-error Allow access
		themeRegistry: aesthetic.themeRegistry,
	};
}

export function setupAesthetic(aesthetic: Aesthetic<any, any>) {
	aesthetic.registerTheme('day', lightTheme);
	aesthetic.registerTheme('night', darkTheme);
}

export function teardownAesthetic(aesthetic: Aesthetic<any, any>) {
	lightTheme.name = '';
	darkTheme.name = '';
	resetAestheticState(aesthetic);
}
