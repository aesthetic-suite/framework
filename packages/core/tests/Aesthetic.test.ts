import directionConverter from '@aesthetic/addon-direction';
import vendorPrefixer from '@aesthetic/addon-vendor';
import { createClientEngine } from '@aesthetic/style';
import { createTestStyleEngine, purgeStyles } from '@aesthetic/style/test';
import { ClassName } from '@aesthetic/types';
import { Aesthetic, RenderResultSheet, StyleSheet } from '../src';
import {
  darkTheme,
  getAestheticState,
  lightTheme,
  setupAesthetic,
  teardownAesthetic,
} from '../src/test';

describe('Aesthetic', () => {
  let aesthetic: Aesthetic;

  beforeEach(() => {
    aesthetic = new Aesthetic();
    aesthetic.configureEngine(createClientEngine());
  });

  afterEach(() => {
    purgeStyles();
    teardownAesthetic(aesthetic);
    document.documentElement.setAttribute('dir', 'ltr');
  });

  it('can subscribe and unsubscribe events', () => {
    const spy = jest.fn();

    aesthetic.subscribe('change:theme', spy);

    expect(getAestheticState(aesthetic).listeners.get('change:theme')?.has(spy)).toBe(true);

    aesthetic.unsubscribe('change:theme', spy);

    expect(getAestheticState(aesthetic).listeners.get('change:theme')?.has(spy)).toBe(false);
  });

  it('can subscribe and unsubscribe events using the return value', () => {
    const spy = jest.fn();

    const unsub = aesthetic.subscribe('change:theme', spy);

    expect(getAestheticState(aesthetic).listeners.get('change:theme')?.has(spy)).toBe(true);

    unsub();

    expect(getAestheticState(aesthetic).listeners.get('change:theme')?.has(spy)).toBe(false);
  });

  describe('changeDirection()', () => {
    beforeEach(() => {
      setupAesthetic(aesthetic);

      // @ts-expect-error
      aesthetic.activeDirection.reset();
    });

    it('sets active direction', () => {
      expect(getAestheticState(aesthetic).activeDirection).toBeUndefined();

      aesthetic.changeDirection('rtl');

      expect(getAestheticState(aesthetic).activeDirection).toBe('rtl');
    });

    it('sets direction on engine', () => {
      expect(aesthetic.getEngine().direction).toBe('ltr');

      aesthetic.changeDirection('rtl');

      expect(aesthetic.getEngine().direction).toBe('rtl');
    });

    it('doesnt run if changing to same name', () => {
      const spy = jest.fn();

      aesthetic.subscribe('change:direction', spy);

      aesthetic.changeDirection('rtl');
      aesthetic.changeDirection('rtl');

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('sets document attribute', () => {
      document.documentElement.dir = 'ltr';

      expect(document.documentElement.dir).toBe('ltr');

      aesthetic.changeDirection('rtl');

      expect(document.documentElement.dir).toBe('rtl');
    });

    it('emits `change:direction` event', () => {
      const spy = jest.fn();

      aesthetic.subscribe('change:direction', spy);

      aesthetic.changeDirection('rtl');

      expect(spy).toHaveBeenCalledWith('rtl');
    });

    it('doesnt emit `change:direction` event if `propagate` is false', () => {
      const spy = jest.fn();

      aesthetic.subscribe('change:direction', spy);

      aesthetic.changeDirection('rtl', false);

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('changeTheme()', () => {
    function createTempSheet() {
      return aesthetic.createThemeStyles(() => ({
        '@root': {
          display: 'block',
          color: 'black',
        },
      }));
    }

    beforeEach(() => {
      setupAesthetic(aesthetic);
    });

    it('sets active theme', () => {
      expect(getAestheticState(aesthetic).activeTheme).toBeUndefined();

      aesthetic.changeTheme('night');

      expect(getAestheticState(aesthetic).activeTheme).toBe('night');
    });

    it('doesnt run if changing to same name', () => {
      const spy = jest.spyOn(aesthetic, 'renderThemeStyles');

      aesthetic.changeTheme('night');
      aesthetic.changeTheme('night');

      expect(spy).toHaveBeenCalledTimes(1);

      spy.mockRestore();
    });

    it('applies root css variables if `rootVariables` is true', () => {
      const spy = jest.spyOn(aesthetic.getEngine(), 'setRootVariables');

      // @ts-expect-error
      aesthetic.options.rootVariables = true;
      aesthetic.changeTheme('night');

      expect(spy).toHaveBeenCalledWith(darkTheme.toVariables());

      spy.mockRestore();
    });

    it('doesnt apply root css variables if `rootVariables` is false', () => {
      const spy = jest.spyOn(aesthetic.getEngine(), 'setRootVariables');

      aesthetic.changeTheme('night');

      expect(spy).not.toHaveBeenCalled();

      spy.mockRestore();
    });

    it('renders theme sheet and sets body class name', () => {
      teardownAesthetic(aesthetic);

      aesthetic.configureEngine(createTestStyleEngine());
      aesthetic.registerTheme('night', darkTheme, createTempSheet());

      aesthetic.changeTheme('night');

      expect(document.body.className).toBe('csehgq8');
    });

    it('emits `change:theme` event', () => {
      const spy = jest.fn();

      aesthetic.subscribe('change:theme', spy);
      aesthetic.changeTheme('night');

      expect(spy).toHaveBeenCalledWith('night', ['csehgq8']);
    });

    it('doesnt emit `change:theme` event if `propagate` is false', () => {
      const spy = jest.fn();

      aesthetic.subscribe('change:theme', spy);
      aesthetic.changeTheme('night', false);

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('configure()', () => {
    it('sets options', () => {
      expect(getAestheticState(aesthetic).options).toEqual({});

      aesthetic.configure({
        defaultUnit: 'em',
        directionConverter,
        vendorPrefixer,
      });

      expect(getAestheticState(aesthetic).options).toEqual({
        defaultUnit: 'em',
        directionConverter,
        vendorPrefixer,
      });
    });

    it('can customize through constructor', () => {
      aesthetic = new Aesthetic({
        defaultUnit: 'em',
        directionConverter,
        vendorPrefixer,
      });

      expect(getAestheticState(aesthetic).options).toEqual({
        defaultUnit: 'em',
        directionConverter,
        vendorPrefixer,
      });
    });
  });

  describe('createComponentStyles()', () => {
    beforeEach(() => {
      setupAesthetic(aesthetic);
    });

    it('returns a `LocalSheet` instance', () => {
      expect(aesthetic.createComponentStyles(() => ({}))).toBeInstanceOf(StyleSheet);
    });
  });

  describe('createThemeStyles()', () => {
    it('returns a `GlobalSheet` instance', () => {
      expect(aesthetic.createThemeStyles(() => ({}))).toBeInstanceOf(StyleSheet);
    });
  });

  describe('generateClassName()', () => {
    const classes: RenderResultSheet<ClassName> = {
      a: { result: 'a', variants: { size_df: 'a_size_df' } },
      b: { result: 'b' },
      c: { result: 'c', variants: { size_md: 'c_size_md', type_red: 'c_size_red' } },
      d: { variants: { size_df: 'd_size_df' } },
      e: { result: 'e' },
      f: { result: 'f', variants: { size_df: 'f_size_df', size_md: 'f_size_md' } },
      g: { result: 'g', variants: { type_red: 'c_size_red' } },
    };

    it('returns class names', () => {
      expect(aesthetic.generateClassName(['a', 'e'], [], classes)).toBe('a e');
    });

    it('returns class names and their variants', () => {
      expect(aesthetic.generateClassName(['a'], ['size_df'], classes)).toBe('a a_size_df');
      expect(aesthetic.generateClassName(['a', 'f'], ['size_df'], classes)).toBe(
        'a a_size_df f f_size_df',
      );
    });

    it('returns nothing for an invalid selector', () => {
      expect(aesthetic.generateClassName(['z'], [], classes)).toBe('');
    });

    it('returns nothing for a valid selector but no class name', () => {
      expect(aesthetic.generateClassName(['d'], [], classes)).toBe('');
    });

    it('returns variants even if theres no base class name', () => {
      expect(aesthetic.generateClassName(['d'], ['size_df'], classes)).toBe('d_size_df');
    });
  });

  describe('getActiveDirection()', () => {
    beforeEach(() => {
      // @ts-expect-error
      aesthetic.activeDirection.reset();
    });

    it('returns the direction defined on the html `dir` attribute', () => {
      const changeSpy = jest.fn();

      aesthetic.subscribe('change:direction', changeSpy);

      document.documentElement.setAttribute('dir', 'rtl');
      document.body.removeAttribute('dir');

      // Reset engine to detect attribute
      aesthetic.configureEngine(createClientEngine());

      expect(aesthetic.getActiveDirection()).toBe('rtl');
      expect(changeSpy).toHaveBeenCalledWith('rtl');
    });

    it('returns the direction defined on the body `dir` attribute', () => {
      document.documentElement.removeAttribute('dir');
      document.body.setAttribute('dir', 'rtl');

      // Reset engine to detect attribute
      aesthetic.configureEngine(createClientEngine());

      expect(aesthetic.getActiveDirection()).toBe('rtl');
    });

    it('returns ltr if no dir found', () => {
      document.documentElement.removeAttribute('dir');
      document.body.removeAttribute('dir');

      // Reset engine to detect attribute
      aesthetic.configureEngine(createClientEngine());

      expect(aesthetic.getActiveDirection()).toBe('ltr');
    });

    it('caches result for subsequent lookups', () => {
      expect(aesthetic.getActiveDirection()).toBe('ltr');

      document.documentElement.setAttribute('dir', 'rtl');

      expect(aesthetic.getActiveDirection()).toBe('ltr');
    });
  });

  describe('getActiveTheme()', () => {
    it('errors if no themes registered', () => {
      expect(() => {
        aesthetic.getActiveTheme();
      }).toThrow('No themes have been registered.');
    });

    it('returns the active theme defined by property', () => {
      const changeSpy = jest.fn();

      aesthetic.subscribe('change:theme', changeSpy);
      setupAesthetic(aesthetic);
      aesthetic.changeTheme('night');

      expect(aesthetic.getActiveTheme()).toBe(darkTheme);
      expect(changeSpy).toHaveBeenCalledWith('night', ['csehgq8']);
    });

    it('returns the preferred theme if no active defined', () => {
      const getSpy = jest.spyOn(getAestheticState(aesthetic).themeRegistry, 'getPreferredTheme');
      const changeSpy = jest.fn();

      aesthetic.subscribe('change:theme', changeSpy);
      setupAesthetic(aesthetic);

      expect(aesthetic.getActiveTheme()).toBe(lightTheme);
      expect(getSpy).toHaveBeenCalled();
      expect(changeSpy).toHaveBeenCalledWith('day', ['c1mbmoiu']);
    });
  });

  describe('getTheme()', () => {
    it('errors if not registered', () => {
      expect(() => {
        aesthetic.getTheme('unknown');
      }).toThrow('Theme "unknown" does not exist. Has it been registered?');
    });

    it('returns the theme defined by name', () => {
      aesthetic.registerTheme('day', lightTheme);

      expect(aesthetic.getTheme('day')).toBe(lightTheme);
    });
  });

  describe('registerDefaultTheme()', () => {
    it('registers a default theme', () => {
      const spy = jest.spyOn(getAestheticState(aesthetic).themeRegistry, 'register');

      aesthetic.registerDefaultTheme('day', lightTheme);

      expect(spy).toHaveBeenCalledWith('day', lightTheme, true);
    });
  });

  describe('registerTheme()', () => {
    it('registers a theme', () => {
      const spy = jest.spyOn(getAestheticState(aesthetic).themeRegistry, 'register');

      aesthetic.registerTheme('day', lightTheme);

      expect(spy).toHaveBeenCalledWith('day', lightTheme, false);
    });

    it('registers an optional sheet', () => {
      const sheet = aesthetic.createThemeStyles(() => ({}));

      aesthetic.registerTheme('day', lightTheme, sheet);

      expect(getAestheticState(aesthetic).globalSheetRegistry.get('day')).toBe(sheet);
    });

    it('errors if sheet is not a `GlobalSheet` instance', () => {
      expect(() => {
        aesthetic.registerTheme(
          'day',
          lightTheme,
          // @ts-expect-error
          123,
        );
      }).toThrow('Rendering theme styles require a `GlobalSheet` instance.');
    });
  });

  describe('getEngine()', () => {
    it('errors if not defined', () => {
      expect(() => {
        // @ts-expect-error
        aesthetic.styleEngine.reset();
        aesthetic.getEngine();
      }).toThrow('No style engine defined. Have you configured one with `configureEngine()`?');
    });

    it('returns a client engine by default', () => {
      expect(aesthetic.getEngine()).toBeDefined();
    });

    it('can define a custom engine by using a global variable', () => {
      const customEngine = createTestStyleEngine();

      global.AESTHETIC_CUSTOM_ENGINE = customEngine;

      expect(aesthetic.getEngine()).toBe(customEngine);
    });
  });

  describe('renderComponentStyles()', () => {
    function createTempSheet() {
      return aesthetic.createComponentStyles(() => ({
        foo: {
          display: 'block',
        },
        bar: {
          color: 'black',

          '@variants': {
            type: {
              red: {
                color: 'red',
              },
            },
          },
        },
        baz: {
          position: 'absolute',
        },
      }));
    }

    beforeEach(() => {
      setupAesthetic(aesthetic);
    });

    it('errors if sheet is not a `LocalSheet` instance', () => {
      expect(() => {
        aesthetic.renderComponentStyles(
          // @ts-expect-error
          123,
        );
      }).toThrow('Rendering component styles require a `LocalSheet` instance.');
    });

    it('returns an empty object if no sheet selectors', () => {
      expect(aesthetic.renderComponentStyles(aesthetic.createComponentStyles(() => ({})))).toEqual(
        {},
      );
    });

    it('renders a sheet and returns an object class name', () => {
      const sheet = createTempSheet();
      const spy = jest.spyOn(sheet, 'render');

      expect(aesthetic.renderComponentStyles(sheet)).toEqual({
        foo: { result: 'a' },
        bar: { result: 'b', variants: { type_red: 'c' }, variantTypes: new Set(['type']) },
        baz: { result: 'd' },
      });
      expect(spy).toHaveBeenCalledWith(aesthetic.getEngine(), lightTheme, {
        direction: expect.any(String),
        vendor: false,
      });
    });
  });

  describe('renderFontFace()', () => {
    it('passes to engine', () => {
      const spy = jest.spyOn(aesthetic.getEngine(), 'renderFontFace');
      const fontFace = {
        fontFamily: 'Roboto',
        src: "url('fonts/Roboto.woff2') format('woff2')",
      };

      aesthetic.renderFontFace(fontFace);

      expect(spy).toHaveBeenCalledWith(fontFace, undefined);

      spy.mockRestore();
    });

    it('supports SSS format', () => {
      const spy = jest.spyOn(aesthetic.getEngine(), 'renderFontFace');

      aesthetic.renderFontFace(
        {
          srcPaths: ['fonts/Roboto.woff2'],
        },
        'Roboto',
      );

      expect(spy).toHaveBeenCalledWith(
        {
          fontFamily: 'Roboto',
          src: "url('fonts/Roboto.woff2') format('woff2')",
        },
        undefined,
      );

      spy.mockRestore();
    });
  });

  describe('renderImport()', () => {
    it('passes to engine', () => {
      const spy = jest.spyOn(aesthetic.getEngine(), 'renderImport');
      const path = 'test.css';

      aesthetic.renderImport(path);

      expect(spy).toHaveBeenCalledWith(path, undefined);

      spy.mockRestore();
    });
  });

  describe('renderKeyframes()', () => {
    it('passes to engine', () => {
      const spy = jest.spyOn(aesthetic.getEngine(), 'renderKeyframes');
      const keyframes = {
        from: { opacity: 0 },
        to: { opacity: 1 },
      };

      aesthetic.renderKeyframes(keyframes);

      expect(spy).toHaveBeenCalledWith(keyframes, undefined, undefined);

      spy.mockRestore();
    });

    it('can pass a name and params', () => {
      const spy = jest.spyOn(aesthetic.getEngine(), 'renderKeyframes');
      const keyframes = {
        from: { opacity: 0 },
        to: { opacity: 1 },
      };

      aesthetic.renderKeyframes(keyframes, 'fade', { direction: 'rtl' });

      expect(spy).toHaveBeenCalledWith(keyframes, 'fade', { direction: 'rtl' });

      spy.mockRestore();
    });
  });

  describe('renderThemeStyles()', () => {
    function createTempSheet() {
      return aesthetic.createThemeStyles(() => ({
        '@root': {
          display: 'block',
          width: '100%',
        },
      }));
    }

    it('renders a sheet and returns class names', () => {
      const sheet = createTempSheet();
      const spy = jest.spyOn(sheet, 'render');

      aesthetic.registerDefaultTheme('day', lightTheme, sheet);

      expect(aesthetic.renderThemeStyles(lightTheme)).toEqual(['c1mbmoiu', 'cemumis']);
      expect(spy).toHaveBeenCalledWith(aesthetic.getEngine(), lightTheme, {
        direction: expect.any(String),
        vendor: false,
      });
    });

    it('renders theme CSS variables as a rule group', () => {
      const sheet = createTempSheet();
      const spy = jest.spyOn(aesthetic.getEngine(), 'renderRuleGrouped');

      aesthetic.registerDefaultTheme('day', lightTheme, sheet);
      aesthetic.renderThemeStyles(lightTheme);

      expect(spy).toHaveBeenCalledWith(
        lightTheme.toVariables(),
        expect.objectContaining({ type: 'global' }),
      );
    });

    it('can customize params with options', () => {
      const sheet = createTempSheet();
      const spy = jest.spyOn(sheet, 'render');

      aesthetic.configure({
        defaultUnit: 'em',
        vendorPrefixer,
      });
      aesthetic.registerDefaultTheme('day', lightTheme, sheet);
      aesthetic.renderThemeStyles(lightTheme, { direction: 'rtl' });

      expect(spy).toHaveBeenCalledWith(aesthetic.getEngine(), lightTheme, {
        direction: 'rtl',
        vendor: true,
      });
    });
  });
});
