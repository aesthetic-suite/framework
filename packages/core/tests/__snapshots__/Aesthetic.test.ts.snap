// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aesthetic createComponentStyles() can utilize mixins 1`] = `".a {font-family: inherit;}.b {font-size: inherit;}.c {font-weight: normal;}.d {word-wrap: break-word;}.e {display: flex;}.f {color: var(--palette-brand-text);}"`;

exports[`Aesthetic createThemeStyles() can utilize mixins 1`] = `"@media (min-width: 45.71em) and (max-width: 68.50em) {.cr4yf6a {font-size: var(--breakpoint-xs-root-text-size); line-height: var(--breakpoint-xs-root-line-height);}}@media (min-width: 68.57em) and (max-width: 91.36em) {.cr4yf6a {font-size: var(--breakpoint-sm-root-text-size); line-height: var(--breakpoint-sm-root-line-height);}}@media (min-width: 91.43em) and (max-width: 114.21em) {.cr4yf6a {font-size: var(--breakpoint-md-root-text-size); line-height: var(--breakpoint-md-root-line-height);}}@media (min-width: 114.29em) and (max-width: 137.07em) {.cr4yf6a {font-size: var(--breakpoint-lg-root-text-size); line-height: var(--breakpoint-lg-root-line-height);}}@media (min-width: 137.14em) {.cr4yf6a {font-size: var(--breakpoint-xl-root-text-size); line-height: var(--breakpoint-xl-root-line-height);}}.cr4yf6a {background-color: var(--palette-neutral-bg-base); color: var(--palette-neutral-text); font-family: var(--typography-font-text); font-size: var(--typography-root-text-size); line-height: var(--typography-root-line-height); text-rendering: optimizeLegibility; text-size-adjust: 100%; margin: 0; padding: 0; -webkit-font-smoothing: antialiased; -moz-osx-font-smoothing: grayscale;}"`;
